File: bodhi-supervisor/brain.py
#!/usr/bin/env python3
"""Brain module - LLM integration for evaluating user focus and generating pushback."""

import os
import json
from pathlib import Path
from typing import Optional
from dotenv import load_dotenv

load_dotenv()

STATE_FILE = Path(__file__).parent / "state.json"


class Brain:
    """Handles LLM communication for focus evaluation and pushback generation."""
    
    def __init__(self, provider: str = "openai", project_path: str = "."):
        self.provider = provider
        self.project_path = project_path
        self.client = None
        self._init_client()
    
    def _init_client(self) -> None:
        """Initialize the LLM client based on available API keys."""
        if self.provider == "openai":
            api_key = os.getenv("OPENAI_API_KEY")
            if api_key:
                from openai import OpenAI
                self.client = OpenAI(api_key=api_key)
                self.model = "gpt-4o-mini"
        elif self.provider == "gemini":
            api_key = os.getenv("GEMINI_API_KEY")
            if api_key:
                import google.generativeai as genai
                genai.configure(api_key=api_key)
                self.client = genai.GenerativeModel("gemini-1.5-flash")
        
        if not self.client:
            raise ValueError(
                f"No API key found for {self.provider}. "
                "Set OPENAI_API_KEY or GEMINI_API_KEY in .env"
            )
    
    def build_prompt(self, goal: str, activity_log: list) -> str:
        """Build the evaluation prompt for the LLM."""
        recent = activity_log[-15:] if activity_log else []
        activity_str = "\n".join(f"- {a}" for a in recent) if recent else "No activity recorded yet."
        
        return f"""You are 'Bodhi', a strict but helpful AI Tech Lead.
The user's declared goal for this session is: "{goal}"
Here is their activity log for the last 15 minutes:
{activity_str}

Your task:
1. Determine if their activity aligns with their goal.
2. If they are ON TRACK, output exactly: "OK"
3. If they are OFF TRACK (distracted, procrastinating, or over-engineering), provide a short, sharp pushback message (max 2 sentences). Call out specifically what they are doing wrong and remind them of the goal."""
    
    def evaluate(self, goal: str, activity_log: list) -> str:
        """Evaluate user activity against goal using LLM."""
        prompt = self.build_prompt(goal, activity_log)
        
        if self.provider == "openai":
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": "You are Bodhi, an assertive AI Tech Lead. Respond only with 'OK' or a pushback message."},
                    {"role": "user", "content": prompt}
                ],
                temperature=0.3,
                max_tokens=100
            )
            return response.choices[0].message.content.strip()
        
        elif self.provider == "gemini":
            response = self.client.generate_content(prompt)
            return response.text.strip()
        
        return "OK"
    
    def suggest_next_task(self, goal: str, git_diff: str, recent_files: list) -> str:
        """Suggest the next high-priority task based on codebase state."""
        # Build context package from git
        import subprocess
        try:
            recent_commits = subprocess.run(
                ["git", "log", "-n", "3", "--oneline"],
                cwd=self.project_path if hasattr(self, 'project_path') else ".",
                capture_output=True, text=True, timeout=5
            ).stdout.strip() or "No recent commits"
        except Exception:
            recent_commits = "Error reading git log"
        
        try:
            git_status = subprocess.run(
                ["git", "status", "--short"],
                cwd=self.project_path if hasattr(self, 'project_path') else ".",
                capture_output=True, text=True, timeout=5
            ).stdout.strip() or "Clean"
        except Exception:
            git_status = "Error reading git status"
        
        prompt = f"""Based on this git state and the user's current goal, what is the single most logical next coding task? Provide it as a 1-sentence instruction.

Goal: "{goal}"

Recent Commits:
{recent_commits}

Current Uncommitted State:
{git_status}

File Changes (stat):
{git_diff[:2000] if git_diff else "No uncommitted changes"}

Recent file activity:
{chr(10).join(f"- {f}" for f in recent_files[-10:])}"""
        
        if self.provider == "openai":
            response = self.client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": "You are Bodhi, an AI Tech Lead. Output one specific next task as a single sentence."},
                    {"role": "user", "content": prompt}
                ],
                temperature=0.4,
                max_tokens=150
            )
            return response.choices[0].message.content.strip()
        
        elif self.provider == "gemini":
            response = self.client.generate_content(prompt)
            return response.text.strip()
        
        return "Continue with current task."


def test_brain() -> None:
    """Test the brain with a mock activity log."""
    # This requires an API key in .env
    try:
        brain = Brain(provider="openai")  # or "gemini"
        test_goal = "Build the user authentication API"
        test_activity = [
            "10:00 - VS Code (auth.py)",
            "10:05 - VS Code (auth.py)",
            "10:10 - Chrome (Stack Overflow - JWT tokens)",
            "10:15 - VS Code (auth.py)",
        ]
        result = brain.evaluate(test_goal, test_activity)
        print(f"Evaluation: {result}")
    except ValueError as e:
        print(f"Skipping test (no API key): {e}")


if __name__ == "__main__":
    test_brain()

File: prototype/src/brain.js
// @ts-check
/** Brain module — LLM integration for evaluating user focus and generating pushback. */

const { app } = require('electron');
const path = require('path');
const fs = require('fs');
const storage = require('./storage');

/** @typedef {'openai' | 'gemini'} Provider */
/** @typedef {{ provider: Provider, model: string, apiKey: string }} LLMConfig */

/** @type {LLMConfig|null} */
let brainConfig = null;

/** @type {any} */
let client = null;

/**
 * Initialize the LLM client from settings.
 * @returns {boolean} true if brain is available
 */
function initBrain() {
  const settings = storage.getSettings();
  const provider = settings.llmProvider || 'openai';
  const apiKey = settings.llmApiKey || '';
  const model = settings.llmModel || (provider === 'openai' ? 'gpt-4o-mini' : 'gemini-1.5-flash');

  if (!apiKey) {
    brainConfig = null;
    client = null;
    return false;
  }

  brainConfig = { provider, model, apiKey };

  if (provider === 'openai') {
    try {
      const OpenAI = require('openai');
      client = new OpenAI({ apiKey });
    } catch (e) {
      console.warn('[brain] openai package not available:', e.message);
      client = null;
      return false;
    }
  } else if (provider === 'gemini') {
    try {
      const { GoogleGenerativeAI } = require('@google/generative-ai');
      const genAI = new GoogleGenerativeAI(apiKey);
      client = genAI.getGenerativeModel({ model });
    } catch (e) {
      console.warn('[brain] @google/generative-ai package not available:', e.message);
      client = null;
      return false;
    }
  }

  return !!client;
}

/**
 * Build the evaluation prompt for the LLM.
 * @param {string} goal
 * @param {string[]} activityLog
 * @returns {string}
 */
function buildPrompt(goal, activityLog) {
  const recent = activityLog.slice(-15);
  const activityStr = recent.length ? recent.map(a => `- ${a}`).join('\n') : 'No activity recorded yet.';

  return `You are 'Bodhi', a strict but helpful AI Tech Lead.
The user's declared goal for this session is: "${goal}"
Here is their activity log for the last 15 minutes:
${activityStr}

Your task:
1. Determine if their activity aligns with their goal.
2. If they are ON TRACK, output exactly: "OK"
3. If they are OFF TRACK (distracted, procrastinating, or over-engineering), provide a short, sharp pushback message (max 2 sentences). Call out specifically what they are doing wrong and remind them of the goal.`;
}

/**
 * Evaluate user activity against goal using LLM.
 * @param {string} goal
 * @param {string[]} activityLog
 * @returns {Promise<string>} 'OK' or pushback message
 */
async function evaluate(goal, activityLog) {
  if (!client || !brainConfig) return 'OK';

  const prompt = buildPrompt(goal, activityLog);

  try {
    if (brainConfig.provider === 'openai') {
      const response = await client.chat.completions.create({
        model: brainConfig.model,
        messages: [
          {
            role: 'system',
            content: 'You are Bodhi, an assertive AI Tech Lead. Respond only with \'OK\' or a pushback message.'
          },
          { role: 'user', content: prompt }
        ],
        temperature: 0.3,
        max_tokens: 100
      });
      return response.choices[0].message.content.trim();
    } else if (brainConfig.provider === 'gemini') {
      const result = await client.generateContent(prompt);
      return result.response.text().trim();
    }
  } catch (e) {
    console.error('[brain] Evaluation error:', e.message);
  }

  return 'OK';
}

/**
 * Suggest the next high-priority task based on codebase state.
 * @param {string} goal
 * @param {string} gitDiff
 * @param {string[]} recentFiles
 * @returns {Promise<string>}
 */
async function suggestNextTask(goal, gitDiff, recentFiles) {
  if (!client || !brainConfig) return 'Brain not available (no API key configured)';

  // Get git context
  const { execSync } = require('child_process');
  const projectPath = app?.getPath('userData') || process.cwd();

  let recentCommits = 'No recent commits';
  let gitStatus = 'Clean';

  try {
    recentCommits = execSync('git log -n 3 --oneline', { cwd: projectPath, timeout: 5000, encoding: 'utf8' }).trim() || 'No recent commits';
  } catch { }

  try {
    gitStatus = execSync('git status --short', { cwd: projectPath, timeout: 5000, encoding: 'utf8' }).trim() || 'Clean';
  } catch { }

  const prompt = `Based on this git state and the user's current goal, what is the single most logical next coding task? Provide it as a 1-sentence instruction.

Goal: "${goal}"

Recent Commits:
${recentCommits}

Current Uncommitted State:
${gitStatus}

File Changes (stat):
${gitDiff?.slice(0, 2000) || 'No uncommitted changes'}

Recent file activity:
${recentFiles.slice(-10).map(f => `- ${f}`).join('\n')}`;

  try {
    if (brainConfig.provider === 'openai') {
      const response = await client.chat.completions.create({
        model: brainConfig.model,
        messages: [
          {
            role: 'system',
            content: 'You are Bodhi, an AI Tech Lead. Output one specific next task as a single sentence.'
          },
          { role: 'user', content: prompt }
        ],
        temperature: 0.4,
        max_tokens: 150
      });
      return response.choices[0].message.content.trim();
    } else if (brainConfig.provider === 'gemini') {
      const result = await client.generateContent(prompt);
      return result.response.text().trim();
    }
  } catch (e) {
    console.error('[brain] Next task suggestion error:', e.message);
  }

  return 'Continue with current task.';
}

/**
 * Check if brain is configured and available.
 * @returns {boolean}
 */
function isAvailable() {
  return !!client && !!brainConfig;
}

/**
 * Get current brain configuration (for settings display).
 * @returns {{ provider: string, model: string, available: boolean }}
 */
function getConfig() {
  return {
    provider: brainConfig?.provider || 'none',
    model: brainConfig?.model || 'none',
    available: isAvailable()
  };
}

module.exports = {
  initBrain,
  evaluate,
  suggestNextTask,
  isAvailable,
  getConfig
};